src/tosyntax/FStar.ToSyntax.Resugar.fs

```diff
@@ -0,0 +1,614 @@
+      mk (A.Var l)
+
+    | Tm_fvar fv -> //a top-level identifier, may be lowercase or upper case
+      //should be A.Var if lowercase
+      //and A.Name if uppercase
+      let a = fv.fv_name.v in
+      let s = if fv.fv_name.v.nsstr.Length=0 then a.str 
+          else a.str.Substring(fv.fv_name.v.nsstr.Length+1) in 
+      if (lid_equals a C.assert_lid 
+            || lid_equals a C.assume_lid
+            || Char.uppercase (String.get s 0) <> String.get s 0) then
+        mk (var s t.pos)
+      else 
+        mk (name s t.pos)
+
+    | Tm_uinst(e, universes) ->
+      let e = resugar_term e in
+      let rec aux e = function
+        | [] -> e
+        | hd::tl ->
+          let univ = resugar_universe hd t.pos in
+          let e = mk (A.App(e, univ, A.UnivApp)) in
+          aux e tl in
+      aux e universes
+
+    | Tm_constant c -> 
+      if is_teff t 
+      then mk (name "Effect" t.pos)
+      else mk (A.Const c)
+        
+    | Tm_type u -> 
+        begin match u with 
+        | U_zero ->  mk (name "Type0" t.pos) 
+        | U_unknown -> mk (name "Type" t.pos)
+        | _ -> 
+            let u = resugar_universe u t.pos in
+            let l = lid_of_path ["Type"] t.pos in
+            mk (A.Construct (l, [(u, A.Nothing)]))
+        end
+        
+    | Tm_abs(xs, body, _) -> //fun x1 .. xn -> body
+      //before inspecting any syntactic form that has binding structure
+      //you must call SS.open_* to replace de Bruijn indexes with names
+      let xs, body = SS.open_term xs body in
+      let patterns = xs |> List.map (fun (x, qual) ->
+        //x.sort contains a type annotation for the bound variable
+        //the pattern `p` below only contains the variable, not the annotation
+        //but, if the user wrote the annotation, then we should record that and print it back
+        //additionally, if we're in verbose mode, e.g., if --print_bound_var_types is set
+        //    then we should print the annotation too
+        let p = A.PatVar(bv_as_unique_ident x, resugar_arg_qual qual) in
+        A.mk_pattern p (S.range_of_bv x)) in
+      let body = resugar_term body in
+      mk (A.Abs(patterns, body))
+
+    | Tm_arrow(xs, body) ->
+      let xs, body = SS.open_comp xs body in
+      let body = resugar_comp body in
+      //let xs, body = SS.open_term xs body in
+      let xs = xs |> List.map (fun (b, qual) -> resugar_bv_as_binder b t.pos) in
+      let rec aux body = function
+        | [] -> body
+        | hd::tl ->
+          let body = mk (A.Product([hd], body)) in
+          aux body tl in
+      aux body xs
+
+    | Tm_refine(x, phi) ->
+      (* bv * term -> binder * term *)
+      let x, phi = SS.open_term [S.mk_binder x] phi in
+      let (b, _) = x.Head in 
+      let b = resugar_bv_as_binder b t.pos in
+      mk (A.Refine(b, resugar_term phi))
+
+    | Tm_app(e, args) ->
+      (* Op("=!=", args) is desugared into Op("~", Op("==") and not resugared back as "=!=" *)
```

Comment:
It's the negation of `==`, not `===`. There doesn't seem to be a token for the negation of `===`.